### PR TITLE
[SYCL] Test reduction for complex numbers.

### PR DESCRIPTION
**Pull request by uditagarwal97**
This PR adds test case for testing identityless reduction for complex numbers.

Corresponding intel/llvm PR: https://github.com/intel/llvm/pull/8425

### CODE REVIEW

File: SYCL/Reduction/reduction_complex_nums.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// only valid for plus operator.
+// TODO: Extend this test case once we support known_identity for std::complex
+// and more operators (apart from plus).
+template <typename T>
+void test_identityless_reduction_for_complex_nums(queue &q) {
+  // Allocate and initialize buffer on the host with all 1's.
+  buffer<std::complex<T>> valuesBuf { 1024 };
+  {
+    host_accessor a { valuesBuf };
+    std::fill(a.begin(), a.end(), std::complex<T>(1));
```
Comment (aelovikov-intel) [inline SYCL/Reduction/reduction_complex_nums.cpp]: I'd suggest to put something into the imaginary part as well. Also, maybe make the values non-uniform.

File: SYCL/Reduction/reduction_complex_nums.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  // Buffer to hold the reduction results.
+  std::complex<T> sumResult = 0;
+  buffer<std::complex<T>> sumBuf { &sumResult, 1 };
+
+  q.submit([&](handler& cgh) {
+    auto inputVals = valuesBuf.template get_access<access_mode::read>(cgh);
```
Comment (aelovikov-intel) [inline SYCL/Reduction/reduction_complex_nums.cpp]:
```suggestion
    accessor inputVals{valuesBuf, cgh, sycl::read_only};
```

**Review by aelovikov-intel — APPROVED** [review]

**Comment by uditagarwal97** [comment]
@steffenlarsen The PR is ready, please merge it. The failures in Jenkins CI are unrelated.

**Comment by steffenlarsen** [comment]
> @steffenlarsen The PR is ready, please merge it. The failures in Jenkins CI are unrelated.

Test seems to be failing on Windows. I believe you need `#include <algorithm>`.

File: SYCL/Reduction/reduction_complex_nums.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  });
+
+  assert(sumBuf.get_host_access()[0] == std::complex<T>(523776, 525824));
+}
+
+int main() {
+  queue q;
+
+  test_identityless_reduction_for_complex_nums<float>(q);
+  test_identityless_reduction_for_complex_nums<double>(q);
```
Comment (steffenlarsen) [inline SYCL/Reduction/reduction_complex_nums.cpp]:
```suggestion
  if (q.get_device().has(aspect::fp64))
    test_identityless_reduction_for_complex_nums<double>(q);
```

**Comment by uditagarwal97** [comment]
@steffenlarsen the test case is passing now for both Windows and Linux. There was a bug in the test case before:

The problem was in the test case, precisely in the following statement:`std::generate(a.begin(), a.end(),
                  [&n] { return std::complex<T>(n, ++n + 1); });`

I was using this statement to initialize the buffer. However, the order of evaluation of the expression <n, ++n + 1> is undefined and that's why for different compilers on Linux and Windows, it was initializing the buffer differently.

**Review by steffenlarsen — APPROVED** [review]
Good find! Looks good. 🚀

**Pull request merged**